### PR TITLE
Shadowlings 4: the unfuckening

### PR DESCRIPTION
**Pull request by ghost**
🆑 Thefastfoodguy
refactor: Refactored shadowlings and thralls to use datum antagonists
balance: Shadowlings now will, while in the darkness, generate shield charges, up to a maximum of 3, which can absorb projectiles!
add: Recent discoveries have found out that the shadowling tumor is capable of secreting large amounts of dopamine, causing eternal glee while serving the shadowlings.
balance: Shadowling Glare now deals more stamina damage
balance: Silicons who have their lights veiled by a shadowling now need to have them fixed by a Roboticist (or really, any guy with tools)
fix: Shadow walk works again
fix: Shadowlings work with the new antag panel
/🆑

### CODE REVIEW

**Comment by ghost** [comment]
Don't worry, i'll test it.

**Comment by Carbonhell** [comment]
as if we'll believe you

**Comment by ghost** [comment]
the git push is taking a long time, so....
capture

**Comment by ghost** [comment]
yes i notice that the special role is Thrall, i fixed that

**Comment by ghost** [comment]
@Carbonhell @GuyonBroadway review please

**Comment by Jujumatic** [comment]
lol banned

**Comment by Carbonhell** [comment]
We wont accept any pr from you till you bring back what you forcefully took from us

**Comment by praisenarsie** [comment]
shut up carbon

anyway slings are removed from our rotation anyway, besides we already have nightmares which we should tell our admins about.

im gonna ask the players whether they want slings back or not

**Comment by ghost** [comment]
Assuming the changes in the OP are actually the changes, it'd be a good PR

**Comment by Jujumatic** [comment]
put a post on forums so people can actually vote on whether they want slings back or not - we had to wait months for it to be removed even though the vast majority people said it was complete ass and should be destroyed

**Comment by Pyko1** [comment]
if there is working code for slings, no reason to not have em back.

Having them on rotation is a different thing tho.

**Comment by Pyko1** [comment]
I mean +1 if it works and is TESTED

**Comment by ghost** [comment]
@praisenarsie I believe they were removed from rotation because of the new antag panel not having them

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		spawn(30)
+			if(!M || QDELETED(M))
+				return
+			M.visible_message("<span class='warning'>[M] suddenly bloats and explodes!</span>", \
+							  "<span class='warning'><b>AAAAAAAAA<font size=3>AAAAAAAAAAAAA</font><font size=4>AAAAAAAAAAAA----</font></span>")
+			playsound(M, 'sound/magic/Disintegrate.ogg', 100, 1)
+			M.gib()
+
+/datum/antagonist/shadowling/greet()
+	to_chat(owner, "<br> <span class='shadowling'><b><font size=3>You are a shadowling!</font></b></span>")
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm]: isn't there a class for bold and large font?

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	to_chat(owner, "<b>Currently, you are disguised as an employee aboard [station_name()]].</b>")
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm]: remove second "]"

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	to_chat(owner, "<b>In your limited state, you have three abilities: Enthrall, Hatch, and Hivemind Commune.</b>")
+	to_chat(owner, "<b>Any other shadowlings are your allies. You must assist them as they shall assist you.</b>")
+	to_chat(owner, "<b>If you are new to shadowling, or want to read about abilities, check the wiki page at https://wiki.hippiestation.com/index.php?title=Shadowling</b><br>")
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm]: redundant `<br>`?

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	to_chat(owner, "<b>You require [SSticker.mode.required_thralls || 15] thralls to ascend.</b><br>")
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm]: same here, unless it's intended to create extra new lines

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		parts += "<span class='greentext big'>The shadowlings have ascended and taken over the station!</span>"
+	else if(!SSticker.mode.shadowling_ascended && check_shadow_death()) //If the shadowlings have ascended, they can not lose the round
+		parts += "<span class='redtext big'>The shadowlings have been killed by the crew!</span>"
+	else if(!SSticker.mode.shadowling_ascended && SSshuttle.emergency.mode >= SHUTTLE_ESCAPE)
+		parts += "<span class='redtext big'>The crew escaped the station before the shadowlings could ascend!</span>"
+	else
+		parts += "<span class='redtext big'>The shadowlings have failed!</span>"
+
+	if(objectives.len)
+		parts += "<b>The shadowlings' objectives were:</b>"
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm:90]: do we need `<br>` to separate this from the the previous line (which tells you if the succeeded, etc.)

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		var/count = 1
+		for(var/datum/objective/objective in objectives)
+			if(objective.check_completion())
+				parts += "<b>Objective #[count]</b>: [objective.explanation_text] <span class='greentext'>Success!</span>"
```
Comment (JohnGinnane) [inline hippiestation/code/datums/antagonists/shadowling.dm:94]: same here

File: hippiestation/code/datums/antagonists/shadowling.dm (hunk 1 above)
Comment (ghost) [reply hippiestation/code/datums/antagonists/shadowling.dm]: bold, yes. large, dunno.

File: hippiestation/code/datums/antagonists/shadowling.dm (hunk 5 above)
Comment (ghost) [reply hippiestation/code/datums/antagonists/shadowling.dm:90]: `parts.Join("<br>")`

File: hippiestation/code/datums/antagonists/shadowling.dm (hunk 6 above)
Comment (ghost) [reply hippiestation/code/datums/antagonists/shadowling.dm:94]: `parts.Join("<br>")`

File: hippiestation/code/datums/antagonists/shadowling.dm (hunk 1 above)
Comment (ghost) [reply hippiestation/code/datums/antagonists/shadowling.dm]: Fixed.
`to_chat(owner, "<br> <span class='shadowling bold big'>You are a shadowling!</span>")`

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm [hunk 7]
```diff
@@ -363,9 +244,38 @@ Made by Xhuis
...
 	shadow_hud.join_hud(shadow_mind.current)
-	set_antag_hud(shadow_mind.current, ((shadow_mind in shadows) ? "shadowling" : "thrall"))
+	set_antag_hud(shadow_mind.current, ((is_shadow(shadow_mind.current)) ? "shadowling" : "thrall"))
 
 /datum/game_mode/proc/update_shadow_icons_removed(datum/mind/shadow_mind)
 	var/datum/atom_hud/antag/shadow_hud = GLOB.huds[ANTAG_HUD_SHADOW]
 	shadow_hud.leave_hud(shadow_mind.current)
 	set_antag_hud(shadow_mind.current, null)
+
+/proc/add_thrall(datum/mind/new_thrall_mind)
```
Comment (JohnGinnane) [inline hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: not sure how I feel about these being root procs. We pass in a mind but I feel like these should still be maybe a `human` proc 

need someone else to look at this

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm [hunk 8]
```diff
@@ -363,9 +244,38 @@ Made by Xhuis
...
+	if(!istype(thrall_mind))
+		return FALSE
+	return thrall_mind.remove_antag_datum(ANTAG_DATUM_THRALL)
+
+/proc/remove_sling(datum/mind/ling_mind)
+	if(!istype(ling_mind))
+		return FALSE
+	return ling_mind.remove_antag_datum(ANTAG_DATUM_SLING)
+
+/proc/is_thrall(var/mob/living/M)
```
Comment (JohnGinnane) [inline hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: definitely should be a living or human proc, doesn't need any args

**Comment by JohnGinnane** [comment]
i definitely recommend a video of some gameplay, because it's a large change and while the code might look okay, nothing beats video

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm (hunk 7 above)
Comment (ghost) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: doesn't really hurt

**Comment by ghost** [comment]
video: http://img.spaceball.cf/2018-05-10%2017-06-46.mp4

**Comment by Uqualm** [comment]
>coding in a game mode that nobody can play because your server gets 10 players at high pop times

OH HIPPPPPPPIE RIPPPPPPPPPPPIE. YOU'RE ALREADY DEAD BUT THEY TRYYYYYYYYYYY TO REVIIIIIIIIIIIIVE. THEY CRYYYYYYYY WE ARE ALIIIIIIIIIIIIIIIIVE. 

TOOOOOOOOOOP KEEEEEEEEK

le pickle

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm (hunk 8 above)
Comment (ghost) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: `is_thrall` can just be a define really

Comment (JohnGinnane) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: oh ye much better idea, I think some `is` procs like `is_silicon` do that

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm [hunk 9]
```diff
@@ -363,9 +244,38 @@ Made by Xhuis
...
+	return istype(M) && M.mind && M.mind.has_antag_datum(ANTAG_DATUM_THRALL)
+
+/proc/is_shadow_or_thrall(var/mob/living/M)
+	return M && (is_thrall(M) || is_shadow(M))
+
+/proc/is_shadow(var/mob/living/M)
```
Comment (JohnGinnane) [inline hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: should definitely use defines for "is_" checks, like so:

image

Comment (JohnGinnane) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: That's in code\__DEFINES\is_helpers.dm

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm (hunk 7 above)
Comment (JohnGinnane) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: This is a bad habit, do like how monkeyize() works, so it's consistent with the rest of the code

`/mob/living/carbon/proc/monkeyize()` is the proc

File: hippiestation/code/game/gamemodes/shadowling/shadowling_abilities.dm [hunk 10]
```diff
@@ -152,9 +152,10 @@
 			qdel(G)
 		for(var/mob/living/H in T.contents)
 			extinguishMob(H)
-		for(var/mob/living/silicon/robot/borgie in T.contents)
-			borgie.update_headlamp(TRUE, 150)
-			borgie.lamp_intensity = 0
+		for(var/mob/living/silicon/robot/borg in T.contents)
```
Comment (JohnGinnane) [inline hippiestation/code/game/gamemodes/shadowling/shadowling_abilities.dm:155]: I wanna see some footage of this from the borg's POV, and especially to see it fixed cause I don't see where in the code this is fixed

File: hippiestation/code/modules/mob/mob_movement.dm [hunk 11]
```diff
@@ -1,8 +1,8 @@
 #define MAX_SW_LUMS 0.2
 #define ALLOW_PULL_THROUGH_WALLS 0
 
-/proc/Is_ShadowWalkable(var/turf/loc)
-	return (loc.get_lumcount()==null || loc.get_lumcount() <= MAX_SW_LUMS)
+/proc/Is_ShadowWalkable(var/turf/T)
```
Comment (JohnGinnane) [inline hippiestation/code/modules/mob/mob_movement.dm]: use a define for this like with the "is_" procs

**Review by JohnGinnane — CHANGES_REQUESTED** [review]
needs changes to be consistent with the rest of the code, also more videos showing everything working correctly

File: hippiestation/code/game/gamemodes/shadowling/shadowling_abilities.dm (hunk 10 above)
Comment (ghost) [reply hippiestation/code/game/gamemodes/shadowling/shadowling_abilities.dm:155]: it's the same as a nightmare disabling their lights

File: hippiestation/code/game/gamemodes/shadowling/shadowling.dm (hunk 9 above)
Comment (ghost) [reply hippiestation/code/game/gamemodes/shadowling/shadowling.dm]: Done

File: code/__DEFINES/~hippie_defines/is_helpers.dm [hunk 12]
```diff
@@ -0,0 +1,4 @@
+#define is_thrall(M) (istype(M, /mob/living) && M.mind && M.mind.has_antag_datum(/datum/antagonist/thrall))
+#define is_shadow(M) (istype(M, /mob/living) && M.mind && M.mind.has_antag_datum(/datum/antagonist/shadowling))
+#define is_shadow_or_thrall(M) (is_thrall(M) || is_shadow(M))
+#define Is_ShadowWalkable(T) (isturf(T) && T.get_lumcount()==null || T.get_rgb_lumcount(r_mul = LIGHT_RED_MULTIPLIER, g_mul = LIGHT_GREEN_MULTIPLIER, b_mul = LIGHT_BLUE_MULTIPLIER) <= MAX_SW_LUMS)
```
Comment (JohnGinnane) [inline code/__DEFINES/~hippie_defines/is_helpers.dm]: nitpicking but make this lower case like the others pls

**Comment by JohnGinnane** [comment]
still need to look at https://github.com/HippieStation/HippieStation/pull/7592/files#r187775980

**Review by JohnGinnane — DISMISSED** [review]
looks good enough

**Comment by ghost** [comment]
@hippiestation/heads

**Comment by praisenarsie** [comment]
nice try... non-maintainer

**Comment by JohnGinnane** [comment]
he right though we need @HippieStation/head input on this and ideally another @HippieStation/maintainers because i don't know gamemodes inside out

**Comment by earthmoonjune** [comment]
im ok with having slings back. should make sure the code is good to go tho. +1 if the code is ok.

**Comment by ghost** [comment]
@t-f-w code is ok

**Comment by Pyko1** [comment]
if the code is good sure +1

**Comment by Carbonhell** [comment]
Closing this unless @steamp0rt is still active and willing to fix what could be broken(mainly that Cleaning/cleaning thing in the dme)

**Comment by ghost** [comment]
I'm active, i'm active, someone should've told me to fix stuff, I'm not paying attention to every single one of my dozens of issues and PRs that nobody comments on!

**Comment by Carbonhell** [comment]
okay i'll review it then

**Comment by ghost** [comment]
Ok den.

File: hippiestation.dme [hunk 13]
```diff
@@ -710,10 +712,10 @@
 #include "code\game\objects\effects\decals\decal.dm"
 #include "code\game\objects\effects\decals\misc.dm"
 #include "code\game\objects\effects\decals\remains.dm"
-#include "code\game\objects\effects\decals\cleanable\aliens.dm"
-#include "code\game\objects\effects\decals\cleanable\humans.dm"
```
Comment (Carbonhell) [inline hippiestation.dme]: fix this shit

**Review by Carbonhell — DISMISSED** [review]
Remove that Cleanable diff and it should be ok

File: hippiestation/code/datums/antagonists/shadowling.dm [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	for(var/obj/effect/proc_holder/spell/S in owner.spell_list)
+		owner.RemoveSpell(S)
+	var/mob/living/M = owner.current
+	if(issilicon(M))
+		M.audible_message("<span class='notice'>[M] lets out a short blip.</span>", \
+						  "<span class='userdanger'>You have been turned into a robot! You are no longer a shadowling! Though you try, you cannot remember anything about your time as one...</span>")
+	else
+		M.visible_message("<span class='big'>[M] screams and contorts!</span>", \
+						  "<span class='userdanger'>THE LIGHT-- YOUR MIND-- <i>BURNS--</i></span>")
+		spawn(30)
```
Comment (Carbonhell) [inline hippiestation/code/datums/antagonists/shadowling.dm]: Hmm

**Comment by Carbonhell** [comment]
Do not merge this, but do reviews, i want 3 reviews before this gets in(after Cleanable and that spawn have been commented)

File: hippiestation/code/datums/antagonists/shadowling.dm (hunk 14 above)
Comment (ghost) [reply hippiestation/code/datums/antagonists/shadowling.dm]: Oof, yikes. Will fix.

**Comment by ghost** [comment]
@Carbonhell fixed what you mentioned in your review

**Comment by Carbonhell** [comment]
It's been two months and i have asked maintainers to codecheck this 5-6 times. This makes me pretty disappointed, I am gonna merge this since the code looks good.

**Review by Carbonhell — APPROVED** [review]

**Comment by ghost** [comment]
:clap: 

also why is my forum account broken, i wanna ban appeal again dammit

**Comment by ghost** [comment]
forums were remade
register a new account

**Comment by ghost** [comment]
I did make a new account. My new one is broken

**Comment by Carbonhell** [comment]
How is it broken? Also might be wise to move this on discord or something, pm me

**Pull request merged**